kornia/feature/sold2/backbones.py

```diff
@@ -0,0 +1,436 @@
+        self.convPb = torch.nn.Conv2d(256, 65, kernel_size=1,
+                                      stride=1, padding=0)
+        self.grid_size = grid_size
+
+    def forward(self, input_features: torch.Tensor) -> torch.Tensor:
+        feat = self.relu(self.convPa(input_features))
+        semi = self.convPb(feat)
+
+        # Convert from semi-dense to dense heatmap
+        junc_prob = softmax(semi, dim=1)
+        junc_pred = pixel_shuffle(junc_prob[:, :-1, :, :], self.grid_size)[:, 0]
+        return junc_pred
+
+
+class PixelShuffleDecoder(nn.Module):
+    """Pixel shuffle decoder used to predict the line heatmap.
+
+    Args:
+        input_feat_dim: channel size of the input features.
+        num_upsample: how many upsamples are performed.
+        output_channel: number of output channels.
+    Returns:
+        the (B, 1, H, W) line heatmap.
+    """
+    def __init__(self,
+                 input_feat_dim: int = 128,
+                 num_upsample: int = 2,
+                 output_channel: int = 2):
+        super().__init__()
+        # Get channel parameters
+        self.channel_conf = self.get_channel_conf(num_upsample)
+
+        # Define the pixel shuffle
+        self.pixshuffle = nn.PixelShuffle(2)
+
+        # Process the feature
+        conv_block_lst = []
+        # The input block
+        conv_block_lst.append(
+            nn.Sequential(
+                nn.Conv2d(input_feat_dim, self.channel_conf[0],
+                          kernel_size=3, stride=1, padding=1),
+                nn.BatchNorm2d(self.channel_conf[0]),
+                nn.ReLU(inplace=True)
+            )
+        )
+
+        # Intermediate block
+        for channel in self.channel_conf[1:-1]:
+            conv_block_lst.append(
+                nn.Sequential(
+                    nn.Conv2d(channel, channel, kernel_size=3,
+                              stride=1, padding=1),
+                    nn.BatchNorm2d(channel),
+                    nn.ReLU(inplace=True)
+                )
+            )
+
+        # Output block
+        conv_block_lst.append(
+            nn.Sequential(nn.Conv2d(self.channel_conf[-1], output_channel,
+                          kernel_size=1, stride=1, padding=0))
+        )
+        self.conv_block_lst = nn.ModuleList(conv_block_lst)
+
+    def get_channel_conf(self, num_upsample: int) -> List[int]:
+        """Get num of channels based on number of upsampling."""
+        if num_upsample == 2:
+            return [256, 64, 16]
+        return [256, 64, 16, 4]
+
+    def forward(self, input_features: torch.Tensor) -> torch.Tensor:
+        # Iterate til output block
+        out = input_features
+        for block in self.conv_block_lst[:-1]:
+            out = block(out)
+            out = self.pixshuffle(out)
+
+        # Output layer
+        out = self.conv_block_lst[-1](out)
+        heatmap = softmax(out, dim=1)[:, 1:, :, :]
+
+        return heatmap
+
+
+class SuperpointDescriptor(nn.Module):
```

Comment:
@ducha-aiki we should check later with other ongoing projects how we could reuse this